### PR TITLE
Fix/bootstrap log msg

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

`controllerStack` pulls & logs k8s events;

## QA steps

- bootstrap into k8s and monitors logs

```bash
$ juju bootstrap microk8s k1
```
## no image cached on the node:
```text
Creating Juju controller "k1" on microk8s/localhost
Creating k8s resources for controller "controller-k1"
Downloading images
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.34 to verify accessibility...

Bootstrap complete, controller "k1" now is available in namespace "controller-k1"

Now you can run
	juju add-model <model-name>
to create a new model to deploy k8s workloads.
```
## image cached on the node already:
```text
Creating Juju controller "k1" on microk8s/localhost
Creating k8s resources for controller "controller-k1"
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.28 to verify accessibility...

Bootstrap complete, controller "k1" now is available in namespace "controller-k1"

Now you can run
	juju add-model <model-name>
to create a new model to deploy k8s workloads.
```

## Documentation changes

None

## Bug reference

None

### CODE REVIEW

File: caas/kubernetes/provider/bootstrap.go [hunk 1]
```diff
@@ -505,8 +507,12 @@ func (c controllerStack) ensureControllerConfigmapAgentConf() error {
 	return c.broker.ensureConfigMap(cm)
 }
 
-func (c controllerStack) createControllerStatefulset() error {
-	numberOfPods := int32(1) // TODO: HA mode!
+func getPodName(pcfg *podcfg.ControllerPodConfig) string {
```
Comment (wallyworld) [inline caas/kubernetes/provider/bootstrap.go]: This package function should be either called getControllerPodName() or preferably live on the ControllerPodConfig receiver. It's too generic at the moment.

File: caas/kubernetes/provider/bootstrap.go [hunk 2]
```diff
@@ -544,10 +550,57 @@ func (c controllerStack) createControllerStatefulset() error {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer w.Kill()
+
+	if err = c.broker.createStatefulSet(spec); err != nil {
+		return errors.Trace(err)
+	}
+	if err = c.syncPodStatus(w); err != nil {
+		return errors.Annotate(err, "fetching pods status for controller")
```
Comment (wallyworld) [inline caas/kubernetes/provider/bootstrap.go]: pods'

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -2231,19 +2233,27 @@ func (k *kubernetesClient) getDeploymentStatus(deployment *apps.Deployment) (str
...
+func (k *kubernetesClient) getStatusFromEvents(parentName string, jujuStatus status.Status) (string, status.Status, error) {
+	events, err := k.getEvents(parentName)
 	if err != nil {
 		return "", "", errors.Trace(err)
 	}
 	var statusMessage string
-	// Take the most recent event.
-	if count := len(eventList.Items); count > 0 {
-		evt := eventList.Items[count-1]
+	if count := len(events); count > 0 {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: can we retain the comment?

File: Gopkg.toml [hunk 4]
```diff
@@ -535,3 +535,7 @@
 [[constraint]]
   name = "k8s.io/apiextensions-apiserver"
   revision = "4d05e43ad98c1edcf2f52291685bd9bbc12fd2cd"
+
+[[constraint]]
```
Comment (wallyworld) [inline Gopkg.toml]: This is a huge dependency to bring in just for the ContainerStarted const.
We can just define our own const and avoid the dependency. Fair enough to bring in the dependency if we use more in the future.

File: caas/kubernetes/provider/bootstrap.go [hunk 5]
```diff
@@ -544,10 +550,57 @@ func (c controllerStack) createControllerStatefulset() error {
...
+	for {
+		select {
+		case <-w.Changes():
+			events, err := c.broker.getEvents(getPodName(c.pcfg))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if checkEvents(events) {
+				// all containers are created.
+				c.ctx.Infof("Pulled images, created %d containers", c.containerCount)
```
Comment (wallyworld) [inline caas/kubernetes/provider/bootstrap.go]: We don't need the "pulled images bit" as if the containers are created,the images must have been pulled.

Immediately after this, we print "Bootstrap agent now started". So "created 2 containers" doesn't really add anything to the user experience.
What we want is to inform the user what is happening between
Creating k8s resources for controller "controller-foo"
and
Bootstrap agent now started

Something like:

Creating k8s resources for controller "controller-foo"
Downloading pod images
Starting controller pod
Bootstrap agent now started

What we would do is print "Starting controller pod" at the start of the syncPodStatus() and block until things are started.
Ideally we'd also check for image downloading to be finished now that we can poll events etc - we log messages to that effect

File: caas/kubernetes/provider/bootstrap.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap.go]: moved to podcfg package

File: caas/kubernetes/provider/k8s.go (hunk 3 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: added back

File: Gopkg.toml (hunk 4 above)
Comment (ycliuhw) [reply Gopkg.toml]: removed this dep, and created a new file named events.go to define const there and moved events related methods there from k8s.go as well

File: caas/kubernetes/provider/bootstrap.go (hunk 5 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap.go]: Now the logs are like 

```
Creating Juju controller "k2" on microk8s/localhost
Creating k8s resources for controller "controller-k2"
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.50 to verify accessibility...

Bootstrap complete, controller "k2" now is available in namespace "controller-k2"

Now you can run
	juju add-model <model-name>
to create a new model to deploy k8s workloads.
```

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[enhance bootstrap message - removing the download image fake message for example.](https://trello.com/c/AEt7iJ2p/716-enhance-bootstrap-message-removing-the-download-image-fake-message-for-example)

**Pull request merged**